src/basic/log.h

```diff
@@ -367,3 +368,38 @@ int log_syntax_invalid_utf8_internal(
 #define DEBUG_LOGGING _unlikely_(log_get_max_level() >= LOG_DEBUG)
 
 void log_setup(void);
+
+typedef struct LogRateLimit {
+        int error;
+        RateLimit ratelimit;
+} LogRateLimit;
+
+void log_dropped_errors(size_t num_dropped_errors, int level);
+
+bool log_error_is_ratelimited(int error, LogRateLimit *log_ratelimit);
+
+#define log_ratelimit_internal(_level, _error, _file, _line, _func, ...)        \
+({                                                                              \
+        int _log_ratelimit_error = (_error);                                    \
+        static LogRateLimit _log_ratelimit = {                                  \
+                .ratelimit = {                                                  \
+                        .interval = (1 * USEC_PER_SEC),                         \
+                        .burst = 1                                              \
+                },                                                              \
+        };                                                                      \
+        size_t num_dropped_errors = ratelimit_num_dropped(&_log_ratelimit.ratelimit); \
+        if (!log_error_is_ratelimited(_error, &_log_ratelimit)) {               \
+                log_dropped_errors(num_dropped_errors, _level);                 \
+                _log_ratelimit_error = log_internal(_level, _error, _file, _line, _func, __VA_ARGS__);  \
+        }                                                                       \
+        _log_ratelimit_error;                                                   \
+})
```

Comment:
This doesn't seem entirely correct yet.

- Right now, the message about similar messages being dropped is logged separately from the original log message. This makes it harder to associate it to the original log message. Instead, what if we append this information to the original log message, so that we'd get something like the following:

```
Failed to write entry: Too Big (Dropped 56 similar messages)
```

You can easily concatenate two static format strings in the macro by putting their variables next to each other without a comma.

- Also, if we get a different error, we reset the rate limit but we don't log about the new error. This means that if the error is fatal and the program exits, we never log about the error. To avoid that, I think when we encounter a new error, we should first log about the ratelimited error, including the amount of similar messages dropped, followed by logging about the new error.

Reply:
- Do you mean that we should keep track of the exact message not just the error type? It's possible for the error message to change between different calls to the macro. In that case, what should we display? What we could do is just add information about the type of error that was dropped: `Error: Too Big (Dropped 56 similar messages)`. WDYT?
- Actually we do log the new message immediately: after `log_dropped_errors`, we call `log_internal` with the new error.

Reply:
Keeping track of the entire message is going to get too complicated. The level and error are the only things we can reasonably keep track of. So I'd apply the same logic to the level as we do to the error, but disregard the rest of the format arguments as we can't reasonably track them.

And thinking about it again, you're right, we already log the new message immediately, so we don't need to log twice. I'd still merge the "dropped similar messages" stuff into the log message though instead of doing a separate log for it.

So to summarize:
- Keep track of the level, reset the rate limit if it changes
- Don't track the full message or any of the other format arguments
- Merge the two log lines into a single log line

Let me know if something is unclear